### PR TITLE
Retry get image on not found error

### PR DESCRIPTION
**Pull request by miminar**
Image may get updated between tag->digest and digest->manifest
resolutions. Retry the operation if such an event occurs.

And skip expired certificate error.

Resolves flake #6232

### CODE REVIEW

**Comment by miminar** [comment]
@mfojtik wanna take a look?

File: test/integration/dockerregistryclient_test.go [hunk 1]
```diff
@@ -97,11 +99,14 @@ func TestRegistryClientConnectPulpRegistry(t *testing.T) {
...
 
 	var image *dockerregistry.Image
-	err = retryWhenUnreachable(t, func() error {
+	err = retryWhenUnreachableOrNotFound(t, func() error {
 		image, err = conn.ImageByTag("library", "rhel", "latest")
 		return err
 	})
 	if err != nil {
+		if strings.Contains(err.Error(), "x509: certificate has expired or is not yet valid") {
+			t.Skip(err)
```
Comment (miminar) [inline test/integration/dockerregistryclient_test.go]: I know, integration tests will log this as:

    ok      TestRegistryClientConnectPulpRegistry

this is just following another precedence on line [228](https://github.com/openshift/origin/pull/6335/files#diff-67b0e91215f1c0300b678ffbb343b89dR228).

**Comment by mfojtik** [comment]
@miminar LGTM

File: test/integration/dockerregistryclient_test.go (hunk 1 above)
Comment (soltysh) [inline test/integration/dockerregistryclient_test.go]: Can you please log it as the one from the referenced line, at least?

**Comment by soltysh** [comment]
Nit, otherwise LGTM.

File: test/integration/dockerregistryclient_test.go (hunk 1 above)
Comment (miminar) [inline test/integration/dockerregistryclient_test.go]: Resolved.

**Comment by deads2k** [comment]
This change isn't as obvious to me as the connection problem retry.  Can you explain the flow in more detail?

**Comment by miminar** [comment]
@deads2k: The [ImageByTag()](https://github.com/miminar/origin/blob/flake-6232/pkg/dockerregistry/client.go#L234) calls [(*v1Repository).getTaggedImage()](https://github.com/miminar/origin/blob/flake-6232/pkg/dockerregistry/client.go#L516) which first retrieves an image id by with given tag (`latest`) and then it tries to fetch an image json with the id using [getImage()](https://github.com/miminar/origin/blob/flake-6232/pkg/dockerregistry/client.go#L656). I don't know much about pulp but I guess we sometimes hit a short interval when the `library/rhel:latest` is being updated and obtained image id is not longer valid - as the flake error suggests:

    the image "6c3a84d798dc449313787502060b6d5b4694d7527d64a7c99ba199e3b2df834e" in repository "library/rhel" was not found and may have been deleted

**Comment by deads2k** [comment]
@miminar so in the case of this particular flake, it's not our code that is choosing the image ID, its against a remote docker repository?

I think I'd be ok with this if you can limit the retry for this error to only operate against repositories that aren't the internal openshift one.  If we start flaking on our own internal docker repository, we need to fail and dig into it, not retry.

**Comment by miminar** [comment]
@deads2k correct. I've just reworked the retry function to accept error patterns to match. The *image not found* can occur only with v1 registries.

File: test/integration/dockerregistryclient_test.go [hunk 2]
```diff
@@ -21,40 +21,53 @@ const (
...
+
+// retryOnErrors invokes given function several times until it succeeds,
+// returns unexpected error or a maximum number of attempts is reached. It
+// should be used to wrap calls to remote registry to prevent test failures
+// because of short-term outages or image updates.
+func retryOnErrors(t *testing.T, errorPatterns []string, f func() error) error {
 	timeout := retryAfter
 	attempt := 0
 	for err := f(); err != nil; err = f() {
 		// Examples of catched error messages:
```
Comment (deads2k) [inline test/integration/dockerregistryclient_test.go]: out of place

File: test/integration/dockerregistryclient_test.go [hunk 3]
```diff
@@ -21,40 +21,53 @@ const (
...
 			}
 		}
 
-		if reachable || attempt >= maxRetryCount {
+		if !match || attempt >= maxRetryCount {
 			return err
 		}
 
-		t.Logf("registry unreachable \"%v\", retry in %s", err, timeout.String())
+		t.Logf("catched error \"%v\", retrying in %s", err, timeout.String())
```
Comment (deads2k) [inline test/integration/dockerregistryclient_test.go]: caught instead of catched.

File: test/integration/dockerregistryclient_test.go [hunk 4]
```diff
@@ -21,40 +21,53 @@ const (
...
+	unreachableErrorPatterns = []string{
+		"dial tcp",
+		"read tcp",
+		"net/http",
+		"could not be reached",
+	}
+
+	// imageNotFoundErrorPatterns will match following error examples:
+	//   the image "..." in repository "..." was not found and may have been deleted
+	imageNotFoundErrorPatterns = []string{
```
Comment (deads2k) [inline test/integration/dockerregistryclient_test.go]: Add to the comment to make it clear that this is only ok to do if you're connecting to a non-internal registry.

**Comment by deads2k** [comment]
two minor comments.  lgtm otherwise

File: test/integration/dockerregistryclient_test.go (hunk 4 above)
Comment (miminar) [inline test/integration/dockerregistryclient_test.go]: done

File: test/integration/dockerregistryclient_test.go (hunk 2 above)
Comment (miminar) [inline test/integration/dockerregistryclient_test.go]: oops, removed

**Comment by soltysh** [comment]
Nits addressed, so 
[merge]

**Pull request merged**